---
title: "Transcriptome Demo"
author: "Tricia"
date: "`r Sys.Date()`"
output: pdf_document
---

# Load required packages (you might have to figure out how to install some of these first...)
```{r message=FALSE}
library(ballgown)
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
```
#this will install ballgown 
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ballgown")
```

# What is this code doing?
```{r}
pheno_data<-data.frame(ids = c("plank01", "plank02", "biofilm01", "biofilm02"),
                       stage = c("planktonic", "planktonic", "biofilm", "biofilm"))

#this makes a data frame titled "pheno_data" using the the data from "plank01", "plank02", "biofilm01", "biofilm02"

```
# create Ballgown object and check transcript number
```{r message=FALSE}
samples.c <- paste('ballgown', pheno_data$ids, sep = '/')
bg<-ballgown(samples = samples.c,meas='all',pData=pheno_data)
bg
```

# <what is this code doing?>
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt

#this code filters the transcripts from the data bg, turning it from a column to a row, then keeps only those rows with variances greater than 1, which is why it went from seeing 5737 transcripts to 5144 transcripts. 
```
# create a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)

#conducts a statistical test and organizes the data frame, Using data set bg_filt, on the transcript level, while taking into account the different stages, and to compute fold changes
```

# choose a transcript to examine more closely (this is a demo, you need to choose another)
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-PA0135",]
```
#Another transcript to examine
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-PA0142",]
```
## what information are you given about this transcript?
#it retrieves the data, stating that it is a transcript, its "id", calculates its fold change count. pval is the pvalue to indicate the significance of expression difference in different stages (biofilm vs planktonic). Qval is q-value indicate controlling a false discovy. Generally lower p-val and qval means more significance in the expression, respective to treatment or condition. 

# <insert brief explanation of this code>
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)

#this filters the data to only include the genes that have a transcript that has a p-val of less than .05, which will show us the ones that have a significance of different expression between planktonic and biofilm 
```
# organize the table <by what metrics is the table being organized?>
```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)

#the table is being organized taking into account multiple things: from sigdiff column titled "pval", ones who have a p-val < 0.05., the column titled "fc" or fold changes, and by descending order of the absolute value of fold count of the transcriptome.
#So we have sorting those with significant difference by ascending p-values and descending fold changes. 

##########################################################
```
# load gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

# pull out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```

# <what is this code doing? hint:compare the above output of head(gene_expression) to this output>
```{r}
colnames(gene_expression) <- c("plank01", "plank02", "biofilm01", "biofilm02")
head(gene_expression)
dim(gene_expression)

#it renamed the column names of gene_expression
```

# load the transcript to gene table and determine the number of transcripts and unique genes
```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))

#there are 5737 transcripts and 4715 unique genes 
```

# plot the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```

## <insert your interpretation of the above graph>

#The above graph is showing the number of transcripts per gene and how frequent the count is. According to the graph, there is a huge number of a gene with just one transcript where the x-limit is about 64 transcripts in one gene. 

# create a plot of how similar the two replicates are for one another. We have two data sets...how can you modify this code in another chunk to create a plot of the other set?
```{r}
x = gene_expression[,"plank01"]
y = gene_expression[,"plank02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (Planktonic, Replicate 1)", ylab="FPKM (Planktonic, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```
#Plot for biofilm set
```{r}
x = gene_expression[,"biofilm01"]
y = gene_expression[,"biofilm02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="purple", cex=0.25,
xlab="FPKM (Biofilm, Replicate 1)", ylab="FPKM (Biofilm, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "blue")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

## What does it mean if the two data sets are similar?
#It means that the 2 data sets and genes in a biofilm and planktonic exhibit similar expression despite different environment 

# create plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="seagreen",
xlab="log2(Fold change) Planktonic vs Biofilm",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```

## interpret the above figure:
#The figure above plotted the differential gene expression values between biofilm and planktonic environment. In this case, the emphasis is on the fold change between planktonic and biofilm with the peak of the bell curve lying between 0 and 20 frequency differential. 

# Plot total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"plank"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"biofilm"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"plank"]+min_nonzero)
y=log2(gene_expression[,"biofilm"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Planktonic FPKM (log2)", ylab="Biofilm FPKM (log2)",
main="Planktonic vs Biofilm FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="magenta", pch=16, cex=0.5)
legend("topleft", "Significant", col="magenta", pch=16)
```

# make a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```

# choose a gene to determine individual expression (pick a different number than I did)
```{r}
ballgown::transcriptNames(bg_filt)[2]
ballgown::geneNames(bg_filt)[2]
```
#a different gene for determining individual espression
```{r}
ballgown::transcriptNames(bg_filt)[4]
ballgown::geneNames(bg_filt)[4]
```
# transform to log2
```{r}
transformed_fpkm <- log2(fpkm[2, ] + 1)
```

# make sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

# plot expression of individual gene
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[2], ' : ', ballgown::transcriptNames(bg_filt)[2]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("lightblue", "salmon"),
        border="darkblue")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```

## interpret the above figure
# The above figure takes into account log transformed fragments per kilobase million of the filtered data sets from biofilm and planktonic and shows the level of expression of gene PA-0002 in both biofilm and planktonic. the points results from the jitter command, which is done to prevent any overlap of data points and prevent the formation of a distinct pattern.